### PR TITLE
Added BallModel

### LINKED ISSUE #47: Improve ball position and velocity prediction
**Is your feature request or enhancement related to a problem? What specifically could be better? Please describe.**
The current implementations for our ball position and velocity estimation are either linear, or decay with an empirically chosen time constant. This does not necessarily reflect the actual behavior of the ball, and should be made more representative.

**Describe the solution you'd like**
We should have a more data-driven way to predict the ball's behavior. This could be accomplished through implementing a model of the ball (ie. how the ball reacts with the world, friction, etc.) that we can update and query to predict behavior.

Another option would be transfer functions derived via experiments. For example, roll the ball on the field several times and log the positional data. Fit a function to the position vs time in order to get a function we can use to predict with (similar with velocity). It would be ideal if it was scripted or quick enough for us to run before games to update our functions based on the field we're playing on (they are not all the same).

There might be other ways to accomplish this as well. Do some research on prediction and don't feel restricted by what I've mentioned above.

**Additional context**
Rules regarding dribblers are still changing, so do *not* consider ball spin in any models for now

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
* Added BallModel virtual class
* Added LinearBallModel for a simple sliding/rolling friction model
* Added tests
* Integrated into intercept
NOTE: sensorfusion will need to update the ball model appropriate as it is currently default constructed to match old behaviour

### Testing Done
added unit tests

### Resolved Issues
part of #1522
resolves #47

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/world/ball_model/linear_ball_model.h [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        const BallState& initial_ball_state, double seconds_in_future,
+        const FrictionParameters& friction_parameters);
+
+    /**
+     * Calculates the future ball state assuming constant acceleration opposing the
+     * velocity of the ball
+     *
+     * @param initial_position initial_position of the ball
+     * @param initial_velocity initial_velocity of the ball
+     * @param friction_acceleration The magnitude of the acceleration due to friction
```
Comment (kelvinkoon) [inline src/software/world/ball_model/linear_ball_model.h]: Nit: None of the other javadoc comments begin capitalized except this one.

**Review by kelvinkoon — DISMISSED** [review]
LGTM, just needs CI.

File: src/software/world/ball_model/linear_ball_model.h (hunk 1 above)
Comment (jonathanlew) [reply src/software/world/ball_model/linear_ball_model.h]: fixed

**Review by kelvinkoon — DISMISSED** [review]
:soccer:

File: src/software/world/ball.cpp [hunk 2]
```diff
@@ -30,14 +35,19 @@ TimestampedBallState Ball::currentState() const
...
+
 void Ball::updateState(const TimestampedBallState &new_state)
 {
     if (!states_.empty() && new_state.timestamp() < lastUpdateTimestamp())
     {
         throw std::invalid_argument(
             "Error: Trying to update ball state using a state older then the current state");
     }
-
+    ball_model_ = std::make_shared<LinearBallModel>(LinearBallModel(new_state.state()));
```
Comment (akhilveeraghanta) [inline src/software/world/ball.cpp]: do we construct this every-time to avoid having an `updateModel` function be part of the interface?

File: src/software/world/ball_model/linear_ball_model.cpp [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                    friction_parameters.rolling_friction_acceleration,
+                    seconds_to_spend_rolling);
+            }
+            else
+            {
+                return calculateFutureBallState(
+                    initial_rolling_state,
+                    friction_parameters.rolling_friction_acceleration,
+                    seconds_until_stopped);
+            }
```
Comment (akhilveeraghanta) [inline src/software/world/ball_model/linear_ball_model.cpp]: What do we need this case for? When will the`seconds_to_spend_rolling` be longer than the time it takes to stop due to the `rolling_friction`?

**Review by akhilveeraghanta — COMMENTED** [review]
This is really neat ⚽ , just a few comments

File: src/software/new_geom/util/physics.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ * @param initial_position The initial position of the object
+ * @param initial_velocity The initial velocity of the object in position
+ * @param acceleration The constant acceleration applied to the object over time
+ * @param time_in_future The time in the future to estimate position
+ *
+ * @return future position of the object
+ */
+Point calculateFuturePosition(const Point &initial_position,
+                              const Vector &intial_velocity, const Vector &acceleration,
+                              double time_in_future);
```
Comment (garethellis0) [inline src/software/new_geom/util/physics.h]: Physics isn't really a logical subset of geometry. I think this should probably be under `src/software` (not  _huge_ fan of adding more stuff at that level, but I think it's most logical place), or alternatively, if this is only used in the linear ball model, we could just put it there (though it would be trickier to test).

File: src/software/new_geom/util/physics.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+ * initial_position)/(units of time_in_future)^2
+ *
+ * @param initial_position The initial position of the object
+ * @param initial_velocity The initial velocity of the object in position
+ * @param acceleration The constant acceleration applied to the object over time
+ * @param time_in_future The time in the future to estimate position
+ *
+ * @return future position of the object
+ */
+Point calculateFuturePosition(const Point &initial_position,
```
Comment (garethellis0) [inline src/software/new_geom/util/physics.h]: Could we rename to `calculateFuturePositionWithConstantAcceleration`? I know it's a bit wordy, but I think it's a _really_ important point to be explicit about the constant acceleration assumption we're making here. 

Similarly for `calculateFutureVelocity` below.

File: src/software/new_geom/util/physics_test.cpp [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    Point expected_p(19.25, -0.25);
+    Vector expected_v(11.5, 3);
+
+    EXPECT_TRUE(TestUtil::equalWithinTolerance(
+        expected_p,
+        calculateFuturePosition(initial_position, initial_velocity, acceleration, 2.5),
+        0.001));
+    EXPECT_TRUE(TestUtil::equalWithinTolerance(
+        expected_v, calculateFutureVelocity(initial_velocity, acceleration, 2.5), 0.001));
+}
```
Comment (garethellis0) [inline src/software/new_geom/util/physics_test.cpp]: We probably want different tests for the different functions, no? I know it's a bit of duplicated code, but I think it helps make the responsibility of each test clearer, and is just generally more consistent with what we have elsewhere.

File: src/software/world/ball.cpp [hunk 7]
```diff
@@ -15,7 +18,9 @@ Ball::Ball(const Point &position, const Vector &velocity, const Timestamp &times
 }
```
Comment (garethellis0) [inline src/software/world/ball.cpp:10]: Could we have this constructor delegate to the other constructor to avoid duplicating code?

File: src/software/world/ball.h [hunk 8]
```diff
@@ -38,9 +38,18 @@ class Ball final
...
+     * @return TimestampedBallState
      */
     TimestampedBallState currentState() const;
 
+    /**
+     * Returns the ball model to predict future states
+     *
+     * @return ball model
+     */
+    const std::shared_ptr<BallModel> &getBallModel() const;
```
Comment (garethellis0) [inline src/software/world/ball.h]: See comment above, I don't think we should be publicly exposing the ball model.

File: src/software/world/ball_model/linear_ball_model.h [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+#include <optional>
+
+#include "software/world/ball_model/ball_model.h"
+
+/**
+ * LinearBallModel is a BallModel that makes the following assumptions about the ball:
+ * 1. no initial spin
+ * 2. slides with a constant sliding friction above a certain speed threshold
+ * 3. rolls with a constant rolling friction below a certain speed threshold
+ */
```
Comment (garethellis0) [inline src/software/world/ball_model/linear_ball_model.h]: These assumptions result in a non-linear model, perhaps `SlideThenRollBallModel` or `SlidingAndRollingBallModel` might be a better name for this class? Or even better, just forget about sliding and rolling, this could just be `TwoFrictionStageBallModel` since we really aren't even doing anything specific to sliding or rolling, we're just applying two different friction coeff's depending on the speed.

Comment (garethellis0) [reply src/software/world/ball_model/linear_ball_model.h]: In order to not block this PR, it might be a good idea to just replace the implementation of this class with a linear, single friction coeff model, then add this more complex model in a future PR?

File: src/software/world/ball_model/linear_ball_model.h [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+{
+   public:
+    struct FrictionParameters
+    {
+        // acceleration opposing the direction of travel of a rolling ball
+        double rolling_friction_acceleration;
+        // acceleration opposing the direction of travel of a sliding ball
+        double sliding_friction_acceleration;
+        // threshold above which the ball slides and below which the ball rolls
+        double rolling_sliding_speed_threshold;
```
Comment (garethellis0) [inline src/software/world/ball_model/linear_ball_model.h]: units on all the members here?

File: src/software/world/ball_model/linear_ball_model.cpp [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+                                initial_ball_state_.velocity() * seconds_in_future;
+        return BallState(future_position, initial_ball_state_.velocity());
+    }
+}
+
+BallState LinearBallModel::applyLinearFrictionModel(
+    const BallState &initial_ball_state, double seconds_in_future,
+    const FrictionParameters &friction_parameters)
+{
+    double initial_velocity_length = initial_ball_state.velocity().length();
```
Comment (garethellis0) [inline src/software/world/ball_model/linear_ball_model.cpp]: I think we can shrink this function a bit by replacing the branches with `min/max` functions. Here's a (very much untested) snippet that I think could work:
```
    const double initial_speed_m_per_s = initial_ball_state.velocity().length();

    // Figure out how long the ball will roll/slide, if at all
    const double max_sliding_duration_secs =
        (initial_speed_m_per_s - friction_parameters.rolling_sliding_speed_threshold) /
        friction_parameters.sliding_friction_acceleration;
    const double sliding_duration_secs =
        std::max(0, std::min(seconds_in_future, max_sliding_duration_secs));
    const double max_rolling_duration_secs =
        std::min(initial_speed_m_per_s,
                 friction_parameters.rolling_sliding_speed_threshold) /
        friction_parameters.rolling_friction_acceleration;
    const double rolling_duration_secs =
        std::max(0, std::min(seconds_in_future - sliding_duration_secs, max_rolling_duration_secs));

    // Figure out where the ball is after the two stages
    BallState ball_state_after_sliding = calculateFutureBallState(
        initial_ball_state, friction_parameters.sliding_friction_acceleration,
        sliding_duration_secs);
    BallState ball_state_after_rolling = calculateFutureBallState(
        ball_state_after_sliding, friction_parameters.rolling_friction_acceleration,
        rolling_duration_secs);

    return ball_state_after_rolling;
```

File: src/software/world/ball_model/linear_ball_model.h (hunk 10 above)
Comment (garethellis0) [reply src/software/world/ball_model/linear_ball_model.h]: Having a struct here feels a tad excessive, I don't think a 4-argument constructor is that ridiculous?

File: src/software/world/ball_model/linear_ball_model.cpp [hunk 12]
```diff
@@ -0,0 +1,113 @@
+#include "software/world/ball_model/linear_ball_model.h"
+
+#include "software/new_geom/util/physics.h"
+
+LinearBallModel::LinearBallModel(BallState initial_ball_state,
+                                 std::optional<FrictionParameters> friction_parameters)
+    : initial_ball_state_(initial_ball_state),
+      friction_parameters_(initFrictionParameters(friction_parameters))
+{
```
Comment (garethellis0) [inline src/software/world/ball_model/linear_ball_model.cpp]: Could we check if the friction parameters are valid and throw an exception if so here? I think it's preferable that we explicitly fail with invalid friction parameters, rather then construct (potentially very invalid) friction parameters by taking the absolute value and carrying on.

**Review by garethellis0 — CHANGES_REQUESTED** [review]
I am extremely curious to know if just having a simple friction model demonstrably improves interception in Sim. :thinking: 

Looking forwards to playing a bit with this!

File: src/software/world/ball.cpp (hunk 2 above)
Comment (jonathanlew) [reply src/software/world/ball.cpp]: the goal is that this function goes away in the world rearch

File: src/software/world/ball_model/linear_ball_model.cpp (hunk 3 above)
Comment (jonathanlew) [reply src/software/world/ball_model/linear_ball_model.cpp]: this is so the ball doesn't keep roll even after friction should've made it stop

File: src/software/ai/evaluation/intercept.cpp [hunk 13]
```diff
@@ -4,9 +4,12 @@
...
 #include "software/optimization/gradient_descent_optimizer.h"
 
-std::optional<std::pair<Point, Duration>> findBestInterceptForBall(
-    TimestampedBallState ball, Field field, TimestampedRobotState robot)
+std::optional<std::pair<Point, Duration>> findBestInterceptForBall(const Ball &ball,
+                                                                   const Field &field,
+                                                                   const Robot &robot)
 {
+    TimestampedBallState ball_state               = ball.currentState();
+    auto ball_model                               = ball.getBallModel();
```
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/ai/evaluation/intercept.cpp]: The idea is to dependency inject different `BallModel`s into `Ball` from the `SensorFusion` layer and not have to change the `Ball` class at all

File: src/software/new_geom/util/physics.h (hunk 4 above)
Comment (jonathanlew) [reply src/software/new_geom/util/physics.h]: Yeah I thought about it. Since it uses geom types, I thought it belonged there. Thoughts?

File: src/software/world/ball.h (hunk 8 above)
Comment (jonathanlew) [reply src/software/world/ball.h]: We're only exposing a pointer to a virtual ball model. The only thing callers will see is the `estimateFutureState` function. No implementation detail should be leaked

File: src/software/ai/evaluation/intercept.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/ai/evaluation/intercept.cpp]: Makes sense, I don't see how wrapping this prevents us doing that?

File: src/software/new_geom/util/physics.h (hunk 4 above)
Comment (garethellis0) [reply src/software/new_geom/util/physics.h]: By that argument pretty much everything in the repo should be in the geom folder! :joy: - That said, I definitely see what you mean. I still think it doesn't belong under the geometry folder.

File: src/software/world/ball_model/ball_model.h [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+     *
+     * @param second_in_future The Duration into the future at which to predict the
+     * ball's position from when the BallModel was initialized
+     *
+     * @throws std::invalid_argument if the ball is estimating state at a time from the
+     * past
+     *
+     * @return The future state of the ball
+     */
+    virtual BallState estimateFutureState(double seconds_in_future) = 0;
```
Comment (MathewMacDougall) [inline src/software/world/ball_model/ball_model.h]: Should use `Duration` instead of `double` for time-based parameters

File: src/software/world/ball_model/linear_ball_model.h [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+     * Creates a new LinearBallModel with the given friction
+     *
+     * @param initial_ball_state The initial state of the ball
+     * @param friction_parameters The FrictionParameters for prediction
+     */
+    explicit LinearBallModel(
+        BallState initial_ball_state,
+        std::optional<FrictionParameters> friction_parameters = std::nullopt);
+
+    BallState estimateFutureState(double seconds_in_future) override;
```
Comment (MathewMacDougall) [inline src/software/world/ball_model/linear_ball_model.h]: Should use `Duration` instead of `double` for time-based parameters

File: src/software/world/ball_model/linear_ball_model.h [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+     * Applies linear friction model
+     *
+     * @param initial_ball_state initial ball state
+     * @param seconds_in_future number of seconds into the future
+     * @param friction_parameters defines friction for model
+     *
+     * @return future ball state
+     */
+    static BallState applyLinearFrictionModel(
+        const BallState& initial_ball_state, double seconds_in_future,
```
Comment (MathewMacDougall) [inline src/software/world/ball_model/linear_ball_model.h]: Should use `Duration` instead of `double` for time-based parameters

File: src/software/world/ball_model/linear_ball_model.h [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+     * @param initial_position The initial position of the ball
+     * @param initial_velocity The initial velocity of the ball
+     * @param friction_acceleration The magnitude of the acceleration due to friction
+     * @param seconds_in_future number of seconds into the future
+     *
+     * @return future ball state
+     */
+    static BallState calculateFutureBallState(BallState initial_ball_state,
+                                              double friction_acceleration,
+                                              double seconds_in_future);
```
Comment (MathewMacDougall) [inline src/software/world/ball_model/linear_ball_model.h]: Should use `Duration` instead of `double` for time-based parameters

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/software/new_geom/util/physics.h (hunk 4 above)
Comment (garethellis0) [reply src/software/new_geom/util/physics.h]: Hmm, I don't see any reason why something like `physics ` should be in the `util` folder but other stuff like `geom` has it's own top level folder?

Comment (MathewMacDougall) [reply src/software/new_geom/util/physics.h]: Yea fair enough, top-level `physics` seems most reasonable

File: src/software/ai/evaluation/intercept.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/software/ai/evaluation/intercept.cpp]: Yea on second thought I don't think we should use DynamicParams here. They won't be changed on the fly, and there's no need to "synchronize" them with anything else (like other code or a GUI), so the DynamicParameters are just acting like a struct here (grouping values to be passed in).

I'd prefer if we created a `PhysicsParams` struct that got passed into the `PhysicsWorld` and the relevant parts propagated to the ball. (A struct will let us to the same injection for Robot physics later).

[the remark this answers is not in the document]
Comment (EvanMorcom) [reply src/software/ai/evaluation/intercept.cpp]: I have the same opinion here as the last Ball PR.

IMO, a model of the ball is dependent on the Field and Ball. I don't think it makes sense to predict any future Ball states based solely on the internal Ball model.

If we inject a BallModel into Ball, who are we to say that this Field used in this function represents the Field parameters (such as carpet friction), that were used in the BallModel?

I get how this makes sense for convenience, but in terms of logical consistency of what makes a _physical_ model of the ball I don't agree with putting everything in the Ball class.

File: src/software/new_geom/util/physics.cpp [hunk 18]
```diff
@@ -0,0 +1,23 @@
+#include "software/new_geom/util/physics.h"
+
+Point calculateFuturePosition(const Point &initial_position,
+                              const Vector &initial_velocity, const Vector &acceleration,
+                              double time_in_future)
+{
+    // Using second equation of motion from https://physics.info/motion-equations/
+    // ∆s = v0t + ½at2
```
Comment (EvanMorcom) [inline src/software/new_geom/util/physics.cpp]: ✨ 

Might be missing a ^ in your ^2 here

File: src/software/new_geom/util/physics.h (hunk 5 above)
Comment (EvanMorcom) [reply src/software/new_geom/util/physics.h]: I know there was a comment about allowing for polymorphism in a previous PR.

Either/or works for me, but something to keep on mind here. Having a constant acceleration as a parameter implies the prediction will use const-acc IMO

File: src/software/world/ball_model/linear_ball_model.h (hunk 9 above)
Comment (EvanMorcom) [reply src/software/world/ball_model/linear_ball_model.h]: This 2-stage friction model might be tricky to tune.

I agree that maybe we should just add a single friction model then create a real ball model using a free body diagram

**Review by EvanMorcom — COMMENTED** [review]
:fire: Definitely sweet to be improving our ball model.

My only gripe is that I don't see the benefit of having a virtual BallModel class for creating many different models.

IMO we should create a ball model based on first principals and that would be our one model. The exception being the Simulator might use a different model to keep things ideal.

Edit: I would also like to add that if we ever want to add more functionality to our ball model, like robot collisions (or *walls*) the model will need more information than available to the Ball class.

Edit Edit: We had a long discussion at the house. Although I don't agree with the Ball containing BallModel I do see where the desire for this implementation comes from. I'm OK merging as-is 👍

**Comment by EvanMorcom** [comment]
I'd be happy if we implemented the following ball model: https://physics.stackexchange.com/questions/114191/help-with-a-solid-sphere-sliding-and-then-rolling

Using a coefficient of dynamic friction up until the point of sliding, then a coefficient of rolling friction after that. This should be relatively easy to implement with what you've got right now.

Where the BallModel belongs I would still like to discuss though.

**Comment by jonathanlew** [comment]
> 
> 
> I'd be happy if we implemented the following ball model: https://physics.stackexchange.com/questions/114191/help-with-a-solid-sphere-sliding-and-then-rolling
> 
> Using a coefficient of dynamic friction up until the point of sliding, then a coefficient of rolling friction after that. This should be relatively easy to implement with what you've got right now.
> 
> Where the BallModel belongs I would still like to discuss though.

I like the physics stackexchange link. I agree it should be easy to integrate into the existing structure. 

The question of where the BallModel belongs is best resolved in a call, so I'll put this PR on hold until our next meeting

File: src/software/new_geom/util/physics.cpp [hunk 19]
```diff
@@ -0,0 +1,23 @@
+#include "software/new_geom/util/physics.h"
+
+Point calculateFuturePosition(const Point &initial_position,
+                              const Vector &initial_velocity, const Vector &acceleration,
+                              double time_in_future)
```
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/new_geom/util/physics.cpp]: The thought here is that all of these params are unitless with the assumption that the units make sense. See the comment in physics.h

File: src/software/new_geom/util/physics.h (hunk 5 above)
Comment (jonathanlew) [reply src/software/new_geom/util/physics.h]: I have renamed the param

File: src/software/new_geom/util/physics.cpp (hunk 18 above)
Comment (jonathanlew) [reply src/software/new_geom/util/physics.cpp]: fixed

File: src/software/world/ball_model/ball_model.h (hunk 14 above)
Comment (jonathanlew) [reply src/software/world/ball_model/ball_model.h]: fixed

File: src/software/new_geom/util/physics_test.cpp (hunk 6 above)
Comment (jonathanlew) [reply src/software/new_geom/util/physics_test.cpp]: separated

File: src/software/world/ball.cpp (hunk 7 above)
Comment (jonathanlew) [reply src/software/world/ball.cpp:10]: done

File: src/software/new_geom/util/physics.h (hunk 4 above)
Comment (jonathanlew) [reply src/software/new_geom/util/physics.h]: moved

File: src/software/world/ball_model/linear_ball_model.cpp (hunk 12 above)
Comment (jonathanlew) [reply src/software/world/ball_model/linear_ball_model.cpp]: fixed

File: src/software/world/ball_model/linear_ball_model.cpp (hunk 11 above)
Comment (jonathanlew) [reply src/software/world/ball_model/linear_ball_model.cpp]: pretty much copy pasted it

File: src/software/world/ball_model/linear_ball_model.h (hunk 15 above)
Comment (jonathanlew) [reply src/software/world/ball_model/linear_ball_model.h]: I think internal functions can safely use seconds

**Comment by jonathanlew** [comment]
> I'd be happy if we implemented the following ball model: https://physics.stackexchange.com/questions/114191/help-with-a-solid-sphere-sliding-and-then-rolling

On second thought, implementing this is out of scope for this PR. In some other PR, we would set up the framework for injecting the parameters to the linear ball model and then calculating good values for them: #1609

File: src/software/new_geom/util/physics.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/software/new_geom/util/physics.cpp]: Ah I see. I still think `Duration` is preferable here because it's a bit more clear it's a time-based value, but it can be somewhat "unitless" as well since it can be constructed with seconds or ms, and get either value as well.

For these functions as well as all the private helpers, I think using Duration instead of `double` will just be a bit more clear and easy to use. Given that our code is pretty much exclusively MKS units I don't see a huge benefit to making this really generic.

@EvanMorcom @garethellis0 @akhilveeraghanta @kelvinkoon ?

File: src/software/world/ball_model/linear_ball_model.h [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+#include "software/world/ball_model/ball_model.h"
+
+/**
+ * LinearBallModel is a BallModel that makes the following assumptions about the ball:
+ * 1. no initial spin
+ * 2. slides with a constant sliding friction above a certain speed threshold
+ * 3. rolls with a constant rolling friction below a certain speed threshold
+ * 4. the transition between sliding and rolling is instantaneous
+ */
+class LinearBallModel final : public BallModel
```
Comment (garethellis0) [inline src/software/world/ball_model/linear_ball_model.h]: Sorry, realized I forgot to comment on this last time. Could we rename this to something like `TwoStageBallModel`, or maybe `TwoStageLinearBallModel`? I would expect a `LinearBallModel` to just have constant decel?

File: src/software/new_geom/util/physics.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/new_geom/util/physics.cpp]: I think using `Duration` makes sense here.

File: src/software/world/ball_model/linear_ball_model.h (hunk 20 above)
Comment (jonathanlew) [reply src/software/world/ball_model/linear_ball_model.h]: done

File: src/software/world/ball_model/linear_ball_model.cpp [hunk 21]
```diff
@@ -0,0 +1,113 @@
+#include "software/world/ball_model/linear_ball_model.h"
+
+#include "software/new_geom/util/physics.h"
+
+LinearBallModel::LinearBallModel(BallState initial_ball_state,
+                                 std::optional<FrictionParameters> friction_parameters)
```
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/world/ball_model/linear_ball_model.cpp]: fixed

File: src/software/new_geom/util/physics.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/new_geom/util/physics.cpp]: ok changed

[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/software/new_geom/util/physics.cpp]: I don't see the change?

[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/new_geom/util/physics.cpp]: whoops changed now

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Review by akhilveeraghanta — APPROVED** [review]
LGTM ⚽

**Review by EvanMorcom — APPROVED** [review]
LGTM 🔥

**Review by MathewMacDougall — APPROVED** [review]
LGTM
